packages/odyssey-react-mui/src/useAutocomplete.tsx

```diff
@@ -0,0 +1,398 @@
+        HasMultipleChoices,
+        undefined,
+        IsCustomValueAllowed
+      >["onChange"]
+    >
+  >(
+    (event, value, reason, details) => {
+      onChangeProp?.(event, value, reason, details);
+    },
+    [onChangeProp],
+  );
+
+  const onInputChange = useCallback<
+    NonNullable<
+      MuiUseAutocompleteProps<
+        OptionType,
+        HasMultipleChoices,
+        undefined,
+        IsCustomValueAllowed
+      >["onInputChange"]
+    >
+  >(
+    (event, value, reason) => {
+      onInputChangeProp?.(event, value, reason);
+    },
+    [onInputChangeProp],
+  );
+
+  const Row = ({
+    data,
+    index,
+    setItemSize,
+    style,
+  }: ListChildComponentProps & { setItemSize: SetItemSize }) => {
+    const rowRef = useRef<HTMLDivElement>(null);
+
+    useEffect(() => {
+      if (rowRef.current) {
+        /**
+         * Checking for child height to workaround a bug where the clientHeight of the row isn't updated correctly
+         * @see here if you need to know more: https://github.com/bvaughn/react-window/issues/582#issuecomment-1883074908
+         */
+        const firstChild = rowRef.current.firstElementChild;
+        const height = firstChild
+          ? firstChild.clientHeight
+          : rowRef.current.clientHeight;
+
+        setItemSize(height);
+      }
+    }, [index, rowRef, setItemSize]);
+
+    const baseOption = data[index];
+    const { key, props } = baseOption;
+
+    /**
+     * react-window calculates the absolute positions of the list items, via an inline style, so
+     * we need to add it to each list item that is being rendered in the viewable list window.
+     * @see here if you need to know more: https://github.com/bvaughn/react-window?tab=readme-ov-file#why-is-my-list-blank-when-i-scroll
+     */
+    const styles = useMemo(
+      () => ({
+        ...style,
+        height: "auto",
+      }),
+      [style],
+    );
+
+    return (
+      <div ref={rowRef}>
+        <li {...props} key={key} style={styles} />
+      </div>
+    );
+  };
+
+  const OuterListboxContext = createContext({});
+
+  const OuterListboxElementType = forwardRef<HTMLDivElement>((props, ref) => {
+    const outerProps = useContext(OuterListboxContext);
+    return <div ref={ref} {...props} {...outerProps} />;
+  });
+
+  const useResetCache = (length: number) => {
+    const resetCacheRef = useRef<VariableSizeList>(null);
+    useEffect(() => {
+      if (resetCacheRef.current) {
+        resetCacheRef.current.resetAfterIndex(0, true);
+      }
+    }, [length]);
+    return resetCacheRef;
+  };
+
+  const ListboxComponent = forwardRef<
+    HTMLDivElement,
+    HTMLAttributes<HTMLElement>
+  >(function (props, ref) {
+    const [listHeight, setListHeight] = useState(0);
+
+    const { children, ...other } = props;
+    const itemData: ReactElement[] = (children as ReactElement[]).flatMap(
+      (item: ReactElement & { children?: ReactElement[] }) =>
+        [item].concat(item.children || []),
+    );
```

Comment:
We shouldn't need to do `children as ReactElement[]`.

If TypeScript isn't seeing it, that's because you need to do `Children.toArray(children).flatMap` instead.

Reply:
I just moved this code. Not sure what the original reasoning was

Reply:
Ok, let's update it to use `Children.toArray` :). That way we're not masking potential "code breaks something in an app" issues with `as`. That's what happened with Side Nav recently; an `as` masked the fact that it would break when given the wrong value.

Reply:
I can't get `Children.toArray` to work quite right. TS isn't liking it